purchase_cancel_qty/models/purchase_order_line.py

```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+# Copyright 2017 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, fields, models, _
+from odoo.addons import decimal_precision as dp
+from odoo.exceptions import ValidationError
+
+PRODUCT_QTY_READONLY_STATES = {
+    'purchase': [('readonly', True)],
+    'done': [('readonly', True)],
+    'cancel': [('readonly', True)],
+}
+
+CANCELLED_QTY_EDITABLE_STATES = {
+    'purchase': [('readonly', False)],
+}
+
+
+class PurchaseOrderLine(models.Model):
+
+    _inherit = 'purchase.order.line'
+
+    product_qty = fields.Float(states=PRODUCT_QTY_READONLY_STATES)
+    ordered_qty = fields.Float(
+        string="Ordered Qty", readonly=True, copy=False,
+        digits=dp.get_precision('Product Unit of Measure'))
+    cancelled_qty = fields.Float(
+        string="Cancelled Qty", readonly=True, copy=False,
+        inverse='_inverse_cancelled_qty',
+        track_visibility='onchange',
+        digits=dp.get_precision('Product Unit of Measure'),
+        states=CANCELLED_QTY_EDITABLE_STATES)
```

Comment:
Ah, one last minor style nitpicking:

```python
product_qty = fields.Float(
    states=PRODUCT_QTY_READONLY_STATES,
)
ordered_qty = fields.Float(
    string="Ordered Qty", 
    readonly=True, 
    copy=False,
    digits=dp.get_precision('Product Unit of Measure'),
)
cancelled_qty = fields.Float(
    string="Cancelled Qty", 
    readonly=True, 
    copy=False,
    inverse='_inverse_cancelled_qty',
    track_visibility='onchange',
    digits=dp.get_precision('Product Unit of Measure'),
    states=CANCELLED_QTY_EDITABLE_STATES,
)
```

Reply:
This style makes it easier to spot all fields, and makes diffs easier to read when adding/modifying fields.